### DISCUSSION TITLE
`$reset` magic variable

### DISCUSSION THREAD

**Discussion by ryangjchandler** [alpinejs/alpine#693] [4. General]
I'm starting a new discussion to get some thoughts on adding a new `$reset` magic variable to Alpine. For those familiar with Livewire, this will essentially behave the same.

When called, you can pass in any number of arguments and each prop will be reset to the initial value (when the component was first created).

```html
<button @click="$reset('propName', 'anotherProp')"></button>
```

**Comment by SimoTod** [comment]
It sounds okay to me. Do you have any use cases where we need it?

**Reply by ryangjchandler** [reply]
Well, my initial use case was a form that when submitted via AJAX needs to reset / clear the inputs - I ended up creating a custom magic variable for now but it was annoying having to assign each variable back to its empty state / initial state.

**Reply by KevinBatdorf** [reply]
@ryangjchandler One way to reset the entire state completely would be to just reinsert the component.

```html
<button @click="$el.replaceWith($el.cloneNode(true))">Reset</button>
```

https://codepen.io/KevinBatdorf/pen/qBZEzoz?editors=1011

**Reply by ryangjchandler** [reply]
Yeah that's fine you need to reset ALL of the state - I've got some of the form state that needs to remain, hence why I didn't do that 😉

**Comment by KevinBatdorf** [comment]
Just brainstorming of course, but instead of a reset, what about having an optional initial state attribute? it could look like this:

```html
<div x-data="{
  name: '',
  email: '',
  send() {}
}"
x-initial-state="{name: 'Kevin', email: ''}">
</div>
```

Then for a reset you could just clone the element, update the `x-initial-state` attribute, then replace. 
edit: [demo](https://codepen.io/KevinBatdorf/pen/wvGavdo)
```html
<button @click="const clone = $el.cloneNode(true);clone.setAttribute('x-initial-state', '{//state goes here}');$el.replaceWith(clone)">Reset</button>
```
It's a bit wordy, but a helper `$reset` method could be available to do this:
```html
<button @click="$reset({//state goes here})">Reset</button>
```

Then Alpine wont need to store the initial state of every component. It would be up to the developer to do so though if they need to.

**Reply by SimoTod** [reply]
Not sure if we need an additional variable. We can just evaluate x-data (which carries the initial state) and set only the variables we need.
Recreating the component would run x-init again which may or may not be what we want (open to discussion)

**Reply by ryangjchandler** [reply]
Yeah that's all I'm doing atm, just re-evaluating `x-data`    . I might make a PR for this today.

**Reply by KevinBatdorf** [reply]
Just adding to the discussion. I don't have an opinion on this too much.

`x-data` might not have the initial state so it would be nice to be able to override it. I often pass in an object in, like:

```js
var componentAState = { name: 'Kevin' }
function componentHandler(state) {
	return Object.assign({
		name: ''
	}, state)
}
// <div x-data="componentHandler(componentAState)"></div>
```

The state for component A might change if they update the name _and press save_, for example. What would reset look like in this situation? Would it still use `componentAState`?

Just for fun, if anyone ends up here looking for something like I described in the initial comment above, here's a demo: https://codepen.io/KevinBatdorf/pen/wvGavdo

**Reply by HugoDF** [reply]
@ryangjchandler would it be possible to create `$reset` in userland? I guess the only tricky point would be re-evaluation of `x-data` with the same logic as `saferEval`

**Reply by ryangjchandler** [reply]
@HugoDF Yeah, I'm throwing a package together at the moment instead of a core implementation. The `$reset` magic variable will return a function that can then do all of the evaluation.

**Reply by ryangjchandler** [reply]
@HugoDF This is a very rough implementation that takes advantage of the `evaluateReturnExpression` method. https://codepen.io/ryangjchandler/pen/zYqGxbq?editors=1010

**Reply by KevinBatdorf** [reply]
@ryangjchandler I didn't know you could do that...! `addMagicProperty` opens up a few ideas I've had that I didn't think I could do. What other secrets do you have? :)

**Reply by ryangjchandler** [reply]
Yeah, it's not going to be documented so be careful using it :) I've got a couple ideas too, it's basically the same as saying `Vue.prototype.$reset` which is handy.

**Comment by rinodrummer** [comment]
Sorry for reopening this discussion, but I found out a really simple solution:

I created a component with a `data()` method - like Vue's Options API - that returns an object with default values and a `reset()` method that gets executed in the `init()` method or whenever someone calls it.

Just like this:

```js
Alpine.data('ResettableComponent', () => ({
    data() {
        return {
            count: 0,
            char: '',
        }
    },
    init() {
        this.reset();
    },
    randomInt(max) {
        return Math.floor(Math.random() * max);
    },
    increment() {
        this.count++;
    },
    write() {
        const chars = [ 'a', 'b', 'c', 'd', 'e' ];

        this.char = chars[];
    },
    reset(...only) {
        let data = this.data();
        
        if (only.length > 0) {
            const deletable = Object.keys(data)
                .filter((prop) => !only.includes(prop));
            
            for (const prop of deletable) {
                data.hasOwnProperty(prop) && delete data[prop];
            }
        }

        Object.assign(this, data);
    }
}));
```

Obviously my `reset()` method has some problems like not working with nested properties in objects (`bar.foo`) or with objects in arrays (`bar.0.foo`) and surely with some other cases, but at least works for properties on the highest level.

_**EDIT:** Now `reset()` method works properly_

**Reply by rinodrummer** [reply]
Ops, `reset()` works on the opposite: resets the properties not present in `only`, but you got the idea
